server/main.go

```diff
@@ -0,0 +1,278 @@
+			Value:       "https://rpc.gochain.io",
+			Usage:       "rpc api url, 'https://rpc.gochain.io'",
+			Destination: &rpcUrl,
+		},
+		cli.StringFlag{
+			Name:        "mongo-url, m",
+			Value:       "127.0.0.1:27017",
+			Usage:       "mongo connection url, '127.0.0.1:27017'",
+			Destination: &mongoUrl,
+		},
+		cli.StringFlag{
+			Name:        "log, l",
+			Value:       "info",
+			Usage:       "loglevel debug/info/warn/fatal, default is Info",
+			Destination: &loglevel,
+		},
+		cli.StringFlag{
+			Name:        "dist, d",
+			Value:       "../dist/explorer/",
+			Usage:       "folder that should be served",
+			Destination: &wwwRoot,
+		},
+	}
+
+	app.Action = func(c *cli.Context) error {
+		level, _ := zerolog.ParseLevel(loglevel)
+		zerolog.SetGlobalLevel(level)
+		backendInstance = backend.NewBackend(mongoUrl, rpcUrl)
+		r := chi.NewRouter()
+		// A good base middleware stack
+		r.Use(middleware.RequestID)
+		r.Use(middleware.RealIP)
+		r.Use(middleware.Logger)
+		r.Use(middleware.Recoverer)
+		cors2 := cors.New(cors.Options{
+			// AllowedOrigins: []string{"https://foo.com"}, // Use this to allow specific origin hosts
+			AllowedOrigins:   []string{"*"},
+			AllowedMethods:   []string{"GET", "POST", "PUT", "DELETE", "OPTIONS"},
+			AllowedHeaders:   []string{"Accept", "Authorization", "Content-Type", "X-CSRF-Token", "Origin"},
+			ExposedHeaders:   []string{"Link"},
+			AllowCredentials: true,
+			MaxAge:           300, // Maximum value not ignored by any of major browsers
+		})
+		r.Use(cors2.Handler)
+		// Set a timeout value on the request context (ctx), that will signal
+		// through ctx.Done() that the request has timed out and further
+		// processing should be stopped.
+		r.Use(middleware.Timeout(60 * time.Second))
+
+		r.Route("/api/stats", func(r chi.Router) {
+			r.Get("/", getCurrentStats)
+		})
+		r.Route("/api/blocks", func(r chi.Router) {
+			r.Get("/", getListBlocks)
+			r.Get("/{num}", getBlock)
+			r.Get("/{num}/transactions", getBlockTransactions)
+		})
+		r.Route("/api/address", func(r chi.Router) {
+			r.Get("/{address}", getAddress)
+			r.Get("/{address}/transactions", getAddressTransactions)
+			r.Get("/{address}/holders", getTokenHolders)
+			r.Get("/{address}/internal_transactions", getInternalTransactions)
+		})
+		r.Route("/api/transaction", func(r chi.Router) {
+			r.Get("/{hash}", getTransaction)
+		})
+
+		r.Route("/api/richlist", func(r chi.Router) {
+			r.Get("/", getRichlist)
+		})
+
+		r.Route("/", func(r chi.Router) {
+			r.Get("/totalSupply", getTotalSupply)
+			r.Get("/circulatingSupply", getCirculating)
+
+			r.Get("/*", staticHandler)
+		})
+
+		http.ListenAndServe(":8080", r)
+		return nil
+	}
+	err := app.Run(os.Args)
+	if err != nil {
+		log.Fatal().Err(err).Msg("Run")
+	}
+
+}
+func getTotalSupply(w http.ResponseWriter, r *http.Request) {
+	totalSupply, _ := backendInstance.TotalSupply()
+	totalSupply = new(big.Int).Div(totalSupply, big.NewInt(1000000000000000000)) //return in GO instead of wei
```

Comment:
These are rounding off to whole $GO. Should use big.Float or even big.Rat instead for better or complete precision.

Reply:
Maybe something like this?
```go
var weiConvert = big.NewRat(1, 1000000000000000000)

func getTotalSupply(w http.ResponseWriter, r *http.Request) {
	totalSupply, _ := backendInstance.TotalSupply()
	total := new(big.Rat).Mul(new(big.Rat).SetInt(totalSupply), weiConvert) // return in GO instead of wei
	w.Write([]byte(total.FloatString(18)))
}

func getCirculating(w http.ResponseWriter, r *http.Request) {
	circulatingSupply, _ := backendInstance.CirculatingSupply()
	circulating := new(big.Rat).Mul(new(big.Rat).SetInt(circulatingSupply), weiConvert) // return in GO instead of wei
	w.Write([]byte(circulating.FloatString(18)))
}
```